x/incentives/keeper/distribute.go

```diff
@@ -9,11 +9,127 @@ import (
+
+			// Divide each coin by remainingEpochs
+			osmoutils.QuoRawMut(coinsToDistribute, remainingEpochs)
+		}
+
+		// Exit early if nothing to distribute.
+		if coinsToDistribute.IsZero() {
+			// Update the group gauge despite zero coins distributed to ensure the filled epochs are updated.
+			if err = k.updateGaugePostDistribute(ctx, *groupGauge, coinsToDistribute); err != nil {
+				return err
+			}
+			continue
+		}
+
+		// We track this for distributing all the remaining amount in the last
+		// gauge without leaving truncation dust.
+		amountDistributed := sdk.NewCoins()
+
+		// Define variables for brevity
+		totalGroupWeight := group.InternalGaugeInfo.TotalWeight
+		gaugeCount := len(group.InternalGaugeInfo.GaugeRecords)
+
+		// Iterate over underlying gauge records in the group.
+		for gaugeIndex, distrRecord := range group.InternalGaugeInfo.GaugeRecords {
+			// Between 0 and 1. to determine the pro-rata share of the total amount to distribute
+			gaugeDistributionRatio := distrRecord.CurrentWeight.ToLegacyDec().Quo(totalGroupWeight.ToLegacyDec())
+
+			// Loop through `coinsToDistribute` and get the amount to distribute to the current gauge
+			// based on the distribution ratio.
+			currentGaugeCoins := osmoutils.MulDec(coinsToDistribute, gaugeDistributionRatio)
+
+			moduleAccount := k.ak.GetModuleAddress(types.ModuleName)
+
+			// For the last gauge, distribute all remaining amounts.
+			// Special case the last gauge to avoid leaving truncation dust in the group gauge
+			// and consume the amounts in-full.
+			if gaugeIndex == gaugeCount-1 {
+				// TODO: module account sends funds to itself. Should we expose API that does not require this?
+				err = k.AddToGaugeRewards(ctx, moduleAccount, coinsToDistribute.Sub(amountDistributed), distrRecord.GaugeId)
+				if err != nil {
+					// We error in this case instead of silently skipping because AddToGaugeRewards should never fail
+					// unless something fundamental has gone wrong.
+					//
+					// Assumption we are making: no gauge in the group outlives the group.
+					return err
+				}
+				break
+			}
+
+			// TODO: module account sends funds to itself. Should we expose API that does not require this?
+			err = k.AddToGaugeRewards(ctx, moduleAccount, currentGaugeCoins, distrRecord.GaugeId)
+			if err != nil {
+				// We error in this case instead of silently skipping because AddToGaugeRewards should never fail
+				// unless something fundamental has gone wrong.
+				//
+				// Assumption we are making: no gauge in the group outlives the group.
+				return err
+			}
+
+			// Update total distribute amount.
+			amountDistributed = amountDistributed.Add(currentGaugeCoins...)
+		}
+
+		// Update total coins distributed and filled epoch of the group gauge.
+		if err = k.updateGaugePostDistribute(ctx, *groupGauge, coinsToDistribute); err != nil {
+			return err
+		}
+
+		// TODO: move group from active to finished if applicable.
```

Comment:
refactor: implement upcoming / active / finished infrastructure for `Group`s similar to gauges